scrapy/utils/conf.py

```diff
@@ -1,36 +1,54 @@
 import os
 import sys
+import warnings
 from operator import itemgetter
 
 import six
 from six.moves.configparser import SafeConfigParser
 
+from scrapy.exceptions import ScrapyDeprecationWarning
+from scrapy.settings import BaseSettings
 from scrapy.utils.deprecate import update_classpath
 
 
-def build_component_list(base, custom, convert=update_classpath):
-    """Compose a component list based on a custom and base dict of components
-    (typically middlewares or extensions), unless custom is already a list, in
-    which case it's returned.
-    """
+def build_component_list(compdict, custom=None, convert=update_classpath):
+    """Compose a component list from a { class: order } dictionary."""
 
     def _check_components(complist):
         if len({convert(c) for c in complist}) != len(complist):
             raise ValueError('Some paths in {!r} convert to the same object, '
                              'please update your settings'.format(complist))
 
-    if isinstance(custom, (list, tuple)):
-        _check_components(custom)
-        return type(custom)(convert(c) for c in custom)
-
     def _map_keys(compdict):
         _check_components(compdict)
-        return {convert(k): v for k, v in six.iteritems(compdict)}
-
-    compdict = _map_keys(base)
-    compdict.update(_map_keys(custom))
-    items = (x for x in six.iteritems(compdict) if x[1] is not None)
-    return [x[0] for x in sorted(items, key=itemgetter(1))]
+        if isinstance(compdict, BaseSettings):
+            compbs = BaseSettings()
+            for k, v in six.iteritems(compdict):
+                prio = compdict.getpriority(k)
+                compbs.set(convert(k), v, priority=prio)
+            return compbs
```

Comment:
> In that case adding them converted into a new BaseSettings shouldn't make any difference since there aren't any clashes, we can stick to `{convert(k): v for k, v in six.iteritems(compdict)}` (there's no need to return a BaseSettings too).

Keeping track of the priorities (i.e. returning `BaseSettings`) was necessary when there were `base` and `custom` BaseSettings objects with different priorities. But when we drop support for the `(base, custom)` signature you're right, returning a simple dict should be fine.

I'll update this and include the edge case of different paths with different priorities